tenants/tenants.go

```diff
@@ -0,0 +1,588 @@
+package tenants
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	stdlog "log"
+	"math"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"regexp"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/go-chi/chi"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/metalmatze/signal/server/signalhttp"
+	"github.com/prometheus/client_golang/prometheus"
+
+	logsv1 "github.com/observatorium/api/api/logs/v1"
+	metricslegacy "github.com/observatorium/api/api/metrics/legacy"
+	metricsv1 "github.com/observatorium/api/api/metrics/v1"
+	"github.com/observatorium/api/authentication"
+	"github.com/observatorium/api/authorization"
+	"github.com/observatorium/api/opa"
+	"github.com/observatorium/api/proxy"
+	"github.com/observatorium/api/ratelimit"
+	"github.com/observatorium/api/rbac"
+)
+
+const grpcDialTimeout = 1 * time.Second
+
+type tenant struct {
+	Name string `json:"name"`
+	ID   string `json:"id"`
+	OIDC *struct {
+		ClientID      string `json:"clientID"`
+		ClientSecret  string `json:"clientSecret"`
+		GroupClaim    string `json:"groupClaim"`
+		IssuerRawCA   []byte `json:"issuerCA"`
+		IssuerCAPath  string `json:"issuerCAPath"`
+		issuerCA      *x509.Certificate
+		IssuerURL     string `json:"issuerURL"`
+		RedirectURL   string `json:"redirectURL"`
+		UsernameClaim string `json:"usernameClaim"`
+	} `json:"oidc"`
+	MTLS *struct {
+		RawCA  []byte `json:"ca"`
+		CAPath string `json:"caPath"`
+		cas    []*x509.Certificate
+	} `json:"mTLS"`
+	OPA *struct {
+		Query      string   `json:"query"`
+		Paths      []string `json:"paths"`
+		URL        string   `json:"url"`
+		authorizer rbac.Authorizer
+	} `json:"opa"`
+	RateLimits []*struct {
+		Endpoint string   `json:"endpoint"`
+		Limit    int      `json:"limit"`
+		Window   duration `json:"window"`
+	} `json:"rateLimits"`
+}
+
+type tenantsConfig struct {
+	Tenants            []*tenant `json:"tenants"`
+	rateLimitClient    *ratelimit.Client
+	authorizer         rbac.Authorizer
+	ins                signalhttp.HandlerInstrumenter
+	reg                *prometheus.Registry
+	retryCounterMetric *prometheus.CounterVec
+	logger             log.Logger
+}
+
+func Register(
```

Comment:
I would really propose returning error here and avoid any logging of errors. This is really messy in this function and nested function. Just error passing will be much clear. Right now we have mix of everything here...